arch/x86/kvm/vmx/pkvm/hyp/pci.c

```diff
@@ -52,6 +52,18 @@ static int pci_mmcfg_write(u64 address, int size, unsigned long value)
 	return 0;
 }
 
+static bool host_vpci_cfg_space_audit_access(struct pkvm_ptdev *ptdev, u64 offset)
+{
+	if (!ptdev || ptdev->shadow_vm_handle == 0)
```

Comment:
Don't we race with ptdev attach/detach here?

Reply:
Yeah this is a race condition. I updated this function to acquire ptdev->lock.

Reply:
So here I should use atomic test of shadow_vm_handle right?

Reply:
I agree that ptdev->lock does not help here yet (and atomic read of ptdev->shadow_vm_handle would not help either), since the issue is not just how to obtain correct ptdev attach status but how to ensure that this status will not change while we try to write BAR (i.e. to prevent security issue with ptdev already attached and BAR write succeeded immediately after, or a correctness issue with ptdev already detached and BAR write failed immediately after).

So yeah, it looks like ptdev->lock should be held until BAR write finishes.

@JasonChenCJ,

>while currently we defer the setup of guest EPT mapping for ptdev BAR untill EPT violation.

It seems to me that we need to change that anyway, e.g. due to security reasons: some secure MMIO may need to be protected from the host even before the secure VM touches this MMIO page first time. So it seems we should do setup of guest EPT for BAR MMIO already at the VM startup (pKVM has all the needed information for that anyway). 

A draft idea of the final implementation:
- ptdev attach/detach is done together with the initial mapping or unmapping of BAR MMIO in the guest and host EPT, under ptdev->lock
- BAR write audit checks if ptdev is attached and if so, updates the guest and host EPT mappings correspondingly, also under ptdev->lock

Makes sense?

Reply:
>and atomic read of ptdev->shadow_vm_handle would not help either

yes, that's why I said it's a loose control here. Ideally, after ptdev attach, the BAR mmio should already be mapped to guest and BAR reprogramming from host is forbid. 

But meanwhile:
- For guest BAR mmio mapping, we rely host vfio do the emulation, we only know gpa for this BAR mmio when guest program vBAR. 
- For host BAR reprogramming, based on our ptdev attach timing, we still see(& need) BAR access from host after it. To be more detail for one important thing, based on @tzhang449 's analysis, the vfio pci device should be inited after pkvm_attach_ptdev (vfio_config_init is done during vfio_pci_core_enable, while vfio_config_init read BAR info for vdev-rbar[])

So, how about we do like below?
1. for BAR write, we hold ptdev->lock all the time (although it does not help for current solution, as we really get vdev->rbar[] after ptdev attach, and we have lock protection for pci cfg data read/write)
2. for BAR write, if the ptdev already assigned, we still allow BAR writing with previous value
3. continue rely vfio do the BAR emulation, which means the mmio EPT mapping of guest will be defer
4. regarding secure device which need remove MMIO page from host, I'd like to make it as a TODO, I suppose we should still allow host to reprogram its BAR config, but forbid its mmio access based on BAR programing value

Reply:
Updated the patch to release the lock after BAR write.